### ISSUE TITLE
NumPy integer indexing compatibility

### ISSUE THREAD

**Issue by 8me** [scikit-hep/awkward#628]
Is it planned to make `awkward` indexing compatible to `numpy` integer types? I encountered this behavior:
```
In [1]: arr = ak.Array([0,1])                                                                                                                                                                  

In [2]: arr[0]                                                                                                                                                                                 
Out[2]: 0

In [3]: arr[np.int64(0)]                                                                                                                                                                       
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-29-cf889115f96f> in <module>
----> 1 arr[np.int64(0)]

~/pyenv/3.7.5/lib/python3.7/site-packages/awkward/highlevel.py in __getitem__(self, where)
    941         have the same dimension as the array being indexed.
    942         """
--> 943         return ak._util.wrap(self._layout[where], self._behavior)
    944 
    945     def __setitem__(self, where, what):

ValueError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`), and integer or boolean arrays (possibly jagged) are valid indices

(https://github.com/scikit-hep/awkward-1.0/blob/1.0.1/src/python/content.cpp#L652)
```
In my case this also occurs within the `awkward` interface, e.g. by doing:
```
In [4]: a = ak.argmax(arr)                                                                                                                                                                     

In [5]: type(a)                                                                                                                                                                                
Out[5]: numpy.int64

In [6]: arr[a]                                                                                                                                                                                 
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-42-bfaae01d2d57> in <module>
----> 1 arr[a]

~/pyenv/3.7.5/lib/python3.7/site-packages/awkward/highlevel.py in __getitem__(self, where)
    941         have the same dimension as the array being indexed.
    942         """
--> 943         return ak._util.wrap(self._layout[where], self._behavior)
    944 
    945     def __setitem__(self, where, what):

ValueError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`), and integer or boolean arrays (possibly jagged) are valid indices

(https://github.com/scikit-hep/awkward-1.0/blob/1.0.1/src/python/content.cpp#L652)
```

**Comment by jpivarski** [comment]
Thanks for finding this! What we ought to do is check to see if an object to be used as a slice has an `__index__` method and call it to get an integer ([PEP 357](https://www.python.org/dev/peps/pep-0357/)). This should happen near the ball-out point mentioned in the error message:

https://github.com/scikit-hep/awkward-1.0/blob/1.0.1/src/python/content.cpp#L652

It's in C++ (the `toslice` function), where we'd use pybind11 methods to check for the existence of a `__index__` and only try to call it if it exists (otherwise, it would be a segfault).

**Comment by jpivarski** [comment]
Correction: the `toslice_part` function.

**Comment by jpivarski** [comment]
I'd call this a bug report, rather than a feature request.

I'll get to it as soon as I can, unless there's someone out there who's interested in a quick project and practical introduction to pybind11.

**Comment by 8me** [comment]
Ok, I was not sure whether this is intentional due to keep it the package light-weight or this really being a bug. I'm not really into the backend of `awkward` so I think I cannot help with resolving the issue, but do you want me to translate my example code into some unit-test?

**Comment by jpivarski** [comment]
Thanks for the offer, but it'll be easy enough to make a unit test from what you've already shown.

This sort of thing would take me about an hour, including the time for running the tests on Azure. I'm just mentioning it as a possible project in case somebody wants a practical introduction to pybind11, but if anybody does, they should speak up so that our implementations don't cross in the mail.

(I'm writing these comments on my phone because I'm on break.)

**Comment by jpivarski** [comment]
I should have done this and included it in 1.0.2rc4! Oh well, I'll do it now.

**Comment by jpivarski** [comment]
Cool: there was already a "FIXME" comment on the offending line of code.

https://github.com/scikit-hep/awkward-1.0/blob/cfac11b07e339ed3f39b27e85c682f4e5a716102/src/python/content.cpp#L411-L415

**Comment by jpivarski** [comment]
This fixes it: 9b521ced4280bc59392e4e3e4c2922029f1fb97c

Anything with a `__index__` method _that does not raise exceptions_ can be used in place of an integer in slices. This includes NumPy integer types like `np.int64(2)` and zero-dimensional NumPy arrays of integers like `np.array([2]).reshape(())`.

**Comment by 8me** [comment]
Cool, thank you!
